crates/doc/src/combine/mod.rs

```diff
@@ -0,0 +1,266 @@
+    match (
+        lhs_doc,
+        lhs_flags & FLAG_REDUCED != 0,
+        rhs_doc,
+        rhs_flags & FLAG_REDUCED != 0,
+    ) {
+        // `rhs_doc` is being combined into `lhs_doc`, which may or may not be fully reduced.
+        (lhs, lhs_reduced, rhs, false) => {
+            let rhs_valid = rhs
+                .validate_ok(validator, schema)
+                .map_err(Error::SchemaError)?
+                .map_err(Error::FailedValidation)?;
+
+            Ok(HeapDoc {
+                root: reduce::reduce(lhs, rhs, rhs_valid, &alloc, lhs_reduced)
+                    .map_err(Error::Reduction)?,
+                flags: if lhs_reduced { FLAG_REDUCED } else { 0 },
+            })
+        }
+        // `rhs_doc` is actually a fully-reduced LHS, which is reduced with `lhs_doc`.
+        (rhs, false, lhs, true) => {
+            let rhs_valid = rhs
+                .validate_ok(validator, schema)
+                .map_err(Error::SchemaError)?
+                .map_err(Error::FailedValidation)?;
+
+            Ok(HeapDoc {
+                root: reduce::reduce(lhs, rhs, rhs_valid, &alloc, true)
+                    .map_err(Error::Reduction)?,
+                flags: FLAG_REDUCED,
+            })
+        }
+        (_lhs, true, rhs, true) => {
+            return Err(Error::AlreadyFullyReduced(serde_json::to_value(&rhs).unwrap()).into())
+        }
+    }
+}
+
+// The bump-allocator threshold after which we'll spill a MemTable to a SpillWriter.
+// We _could_ make this a knob, but empirically using larger values doesn't increase
+// performance in common cases where little reduction is happening, because we're
+// essentially trading a merge-sort of in-memory HeapDocs for a heap merge-sort of
+// ArchivedDocs.
+//
+// If we're writing too many segments -- enough that keeping all of their chunks
+// resident in memory is a problem -- then larger values will decrease the number
+// of segments written.
+//
+// There may be hypothetical use cases that benefit from more in-memory reduction.
+// At the moment, I suspect this would still be pretty marginal.
+const SPILL_THRESHOLD: usize = 8 * (1 << 28) / 10; // 80% of 256MB.
+
+// The chunk target determines the amortization of archiving and
+// compressing documents. We want chunks to be:
+//  - Small, since SpillDrainer must keep one uncompressed chunk in memory
+//    for every segment of the spill file.
+//  - Larger, so that LZ4 can compress it reasonably (we typically see 5x
+//    reduction, 10x is probably the theoretical max).
+//  - Larger, so that filesystem reads and writes are amortized for the
+//    LZ4-compressed.
+//
+// We also want a maximum bound so that SpillDrainer isn't required to
+// hold many large chunks in memory, which could push us over our allocation.
+const CHUNK_TARGET_LEN: usize = 1 << 18; // 256KB.
+const CHUNK_MAX_LEN: usize = 1 << 20; // 1MB.
+
+fn serialize_as_display<T, S>(thing: T, serializer: S) -> Result<S::Ok, S::Error>
```

Comment:
optional nit: could maybe be worth pulling in [serde_with](https://docs.rs/serde_with/latest/serde_with/)